sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala

```diff
@@ -243,6 +243,55 @@ case class MapGroups(
     outputObjAttr: Attribute,
     child: LogicalPlan) extends UnaryNode with ObjectProducer
 
+/** Factory for constructing new `FlatMapGroupsInR` nodes. */
+object FlatMapGroupsInR {
+  def apply(
+      func: Array[Byte],
+      packageNames: Array[Byte],
+      broadcastVars: Array[Broadcast[Object]],
+      schema: StructType,
+      encoder: Expression,
+      keyEncoder: Expression,
+      rowEncoder: ExpressionEncoder[Row],
```

Comment:
Names of the above 3 parameters are quite confusing. The first two are actually deserializer expressions rather than encoders, so I guess `dataDeserializer` and `keyDeserializer` can be better names. The 3rd one is an encoder, but seems that what we actually care about is only its schema.

Basically in most cases you can simply treat encoder as a pair of serializer and deserializer expressions.